Search: Add instant search auto config

gibrown opened the pull request:
For a number of cases we should do some version of auto config on the instant search widgets to improve the initial UX.

Cases this handles:
- no jp search widgets in the overlay or `sidebar-1`
 - if the theme has `sidebar-1` then adds the preconfigured widget and replaces the core search widget if it exists.
 - if the theme has no sidebar then only add it to the overlay
- jp search widget in sidebar but not overlay: copy settings from the sidebar widget into the overlay
- if there is a jp search widget in the overlay: do nothing

The default settings for the widget are look at the post types on the site and registered taxonomies in addition to  tags, categories, and year.

The trigger for the auto config is currently just a hacky query string: `/wp-admin/?jpsearch_autoconfig`. I am not sure what this should get hooked up to. Currently there is no nonce and we probably need to hook it up to something else. It may need to be tied into the onboarding/post-purchase flow though so maybe we should merge this into the feature branch as is for now?

gibrown committed:
[not verified] Auto config of the search widgets.

gibrown committed:
[not verified] Config widgets based on registered post types and taxonomies

jsnmoon commented:
I'm concerned about the implicit nature of our auto-configuration adversely affecting the new customer experience. Could we possibly simplify our approach to the following?

1) If there's a Jetpack Search widget in the sidebar but none in the overlay, duplicate the sidebar widget into the overlay.
2) If there's no Jetpack Search widget in the overlay, add it.
3) Always perform the auto-configuration in the init hook, like `register_jetpack_instant_sidebar`.

Also, I think we could merge this into `master` directly.

gibrown commented:
> If there's a Jetpack Search widget in the sidebar but none in the overlay, duplicate the sidebar widget into the overlay.
> If there's no Jetpack Search widget in the overlay, add it.

It should be doing both of these already (when the query param is present).

> Always perform the auto-configuration in the init hook, like register_jetpack_instant_sidebar.

The register function only puts the info into php global variables. Adding the widgets adds them to the DB can we can't do DB writes on every page load.

> Also, I think we could merge this branch into master directly. There's no other change staged on instant-search-master.

Ah, good point. I think that this should then get combined into #15074 since that is adding an api call for actually triggering the auto config. @AnnaMag was going to take a look at that code to clean up my hackiness and see if that approach makes sense.

So Calypso will make three api calls after  purchase:
- turn on search module
- turn on instant search
- trigger the auto config

Then it will redirect to show the thank you banner in wp-admin.

Does that UX make sense to you @keoshi and @jsnmoon ?

jsnmoon commented:
>So Calypso will make three api calls after purchase:
>- turn on search module
>- turn on instant search
>- trigger the auto config
>
>Then it will redirect to show the thank you banner in wp-admin. @gibrown 

Yup, sounds good! 👍

gibrown commented:
Gah... i need to rebase this onto `add/search-plan` but it is a mess....

keoshi commented:
>So Calypso will make three api calls after purchase:
>- turn on search module
>- turn on instant search
>- trigger the auto config
>
>Then it will redirect to show the thank you banner in wp-admin.

Yep, perfect! 🌟

gibrown committed:
Removed query param for triggering auto config.

gibrown commented:
Rebased this and removed the query param for triggering it so we can use an API. Merging this so we can focus separately on how we want to trigger things.
